Allow empty arrow bodies

wcjohnson opened the pull request:
Fixes https://github.com/wcjohnson/babylon-lightscript/issues/1

wcjohnson committed:
Allow empty arrow bodies

3 squashed commits:

- Allow empty arrow bodies

- Delint

- Change param name

rattrayalex commented on src/plugins/lightscript.js:
```diff
@@ -197,7 +197,8 @@ pp.parseWhiteBlock = function (allowDirectives?, isIfExpression?) {
   this.addExtra(node, "curly", false);
-  if (!node.body.length) this.unexpected(node.start, "Expected an Indent or Statement");
+  if (!isArrowBody && !node.body.length)
+    this.unexpected(node.start, "Expected an Indent or Statement");
```
nit: same line or curly please (surprised this isn't in the eslint rules... hmm)

rattrayalex commented on src/plugins/lightscript.js:
```diff
@@ -168,7 +168,7 @@ pp.parseNumericLiteralMember = function () {
 // c/p parseBlock
 
-pp.parseWhiteBlock = function (allowDirectives?, isIfExpression?) {
+pp.parseWhiteBlock = function (allowDirectives?, isIfExpression?, isArrowBody?) {
```
I don't love adding an extra param here, but I can't think of a cleaner way...

wcjohnson replied:
I think it's necessary due to the way this is used throughout the code. I first tried removing the empty block check entirely, but then the parser allows a lot of malformed if: constructs that it shouldn't. So we need to know if we're in an arrow body or not.

rattrayalex replied:
Yeah. Maybe `allowEmptyBody` as the param name instead? not really impt

rattrayalex reviewed (commented):
I think this looks good... how do you feel about it?

rattrayalex commented on test/fixtures/lightscript/named-arrow-functions/skinny-white-empty/expected.json:
```diff
@@ -0,0 +1,123 @@
+        "body": {
+          "type": "BlockStatement",
+          "start": 5,
+          "end": 7,
```
hmm, odd that the blockstatement has this positioning... looks like it's capturing the arrow... is that happening for other arrow function blocks?

wcjohnson replied:
@rattrayalex  Yes:

https://github.com/wcjohnson/babylon-lightscript/blob/a27269694673bbe26aed026fcdbc23ecc64ca94f/test/fixtures/lightscript/named-arrow-functions/skinny-white/expected.json

`parseWhiteBlock` is calling `startNode` before `eat(tt.Arrow)`, so the arrow's position is mapped into the resulting whiteblock node.

Not sure if that is the intended behavior or not, ultimately, but changing it should probably be out of scope for this PR.

rattrayalex replied:
Makes sense, thanks for checking. Fine to leave for now.

rattrayalex commented:
Cool, I think this looks good... if you're down to change the param name, I'll merge once you've done that + squashed!

wcjohnson commented:
Squashed

rattrayalex commented on src/plugins/lightscript.js:
```diff
@@ -168,7 +168,7 @@ pp.parseNumericLiteralMember = function () {
 // c/p parseBlock
 
-pp.parseWhiteBlock = function (allowDirectives?, isIfExpression?) {
+pp.parseWhiteBlock = function (allowDirectives?, isIfExpression?, allowEmptyBody?) {
```
sorry I didn't catch this before, but instead of doing the param, we can detect by whether there's an arrow in the code itself:

```js
  let allowEmptyBody;
  // must start with colon or arrow
  if (isIfExpression) {
    this.expect(tt.colon);
    if (!this.isLineBreak()) return this.parseMaybeAssign();
  } else if (this.eat(tt.colon)) {
    if (!this.isLineBreak()) return this.parseStatement(false);
  } else if (this.eat(tt.arrow)) {
    allowEmptyBody = true;
    if (!this.isLineBreak()) {
```

Thoughts?

rattrayalex commented on test/fixtures/lightscript/named-arrow-functions/skinny-white-empty/expected.json:
```diff
@@ -0,0 +1,123 @@
+    "sourceType": "script",
+    "body": [
+      {
+        "type": "NamedArrowDeclaration",
```
Sorry I didn't think of this before, but we should probably have a test for non-named arrows, eg:

```
=>
```

perhaps plus the slightly more sane
```
f((x) =>)
````

wcjohnson replied:
These don't work right, basically because `parseWhiteBlock` calls `parseMaybeAssign` here, which expects a non empty expression. I intend to fix this, but odds are you won't like it much.

wcjohnson replied:
https://github.com/wcjohnson/babylon-lightscript/issues/4

wcjohnson replied:
@rattrayalex 

Actually not as bad as I thought:
https://github.com/wcjohnson/babylon-lightscript/commit/997428e2a46ce0eb64dd2799cea0c2544fead504

Let me know if you want this plucked

rattrayalex replied:
Added some comments there, I think... not sure if they went through

wcjohnson replied:
@rattrayalex Saw your comments. Plucked the changeset to this branch, will test and fix here.

wcjohnson committed:
Remove allowEmptyBody param.

wcjohnson commented:
Incidentally, I have compiler fixtures for all this stuff (all passing) at https://github.com/wcjohnson/babel-plugin-lightscript/commit/1d464e9756a23e188e088dfe3004bae2ad51d56d

wcjohnson committed:
Allow empty arrow expressions

- Update `parseWhiteBlock` to only call `parseMaybeAssign` if it seems like the start of an expression.
- Added fixtures

wcjohnson committed:
Add `startsExpr` to Arrow token.

wcjohnson commented:
@rattrayalex 

Changing the arrow token to `startsExpr` worked great, except it broke one Flow fixture in a way that's kind of strange. The fixture break was expected because the `startsExpr` in the token stream changed from false to true, but nonetheless the result is a little strange.

The parse tree being output looks correct to me, including the type annotation data, but the structure of the fixture got changed. See: https://github.com/lightscript/babylon-lightscript/pull/9/commits/e7ff0942579c44dab67a59a9d2d74b5f126f1966

- The return type annotation node got moved before the function body node. Is this because lsc parses flow types in a different order than babylon?

- All the token information got dropped from the regenerated fixture even though it was present in the original, no idea why that happened.

Thoughts?

rattrayalex commented:
The ordering of the nodes doesn't matter, and tokens aren't added to new fixtures these days, so that doesn't matter either. Neither does the identifierName thing. 

What was the incompatible part of the diff? I'm surprised it overwrote it...

wcjohnson commented:
If both those things are expected, then I see no problems with the new fixture.

rattrayalex commented:
I try pretty hard not to modify existing stuff unless it's needed, especially tests. What error is thrown if it's left as it was on master?

wcjohnson commented:
It's expecting `startExpr` to be false when it's true...
```
  index › flow/type annotations/99
  Error: flow/type-annotations/99/actual.js: false !== true (startsExpr/type/7/tokens)
    runTest (test/utils/runFixtureTests.js:164:13)
    Test.fn (test/utils/runFixtureTests.js:25:22)

  index › flow/type annotations/99
  Error: /Users/wcj/dev/lightscript-devkit/packages/babylon-lightscript/test/fixtures/flow/type-annotations/99/actual.js: false !== true (startsExpr/type/7/tokens)
    runTest (test/utils/runFixtureTests.js:164:13)
    Test.fn (test/utils/runFixtureTests.js:107:20)

```

rattrayalex commented:
Ah, gotcha. Let's just delete the tokens from that test then. I'd rather do it another way but this is fine.

wcjohnson committed:
Restore flow fixture; remove tokens

wcjohnson commented:
Restored the old fixture and just removed the token stream; everything passing.

rattrayalex commented on src/tokenizer/types.js:
```diff
@@ -80,7 +80,7 @@ export const types = {
   dot: new TokenType("."),
   question: new TokenType("?", { beforeExpr }),
-  arrow: new TokenType("=>", { beforeExpr }),
+  arrow: new TokenType("=>", { beforeExpr, startsExpr }),
```
ideally this would only happen when `this.hasPlugin("lightscript")`, but I don't think there's an easy way to do that yet. So this is fine for now.

wcjohnson replied:
Yes, unfortunately this is all declarative code, running outside the context of a particular `this` or parser configuration. Would require major changes to the way babylon initializes itself.

rattrayalex commented:
Alright, merging – thanks for this!
